docs/350_REGISTRY.md

```diff
@@ -0,0 +1,55 @@
+# What is a registry?
+
+A `Registry` should be responsible for contract specific behavior and consolidate
+the outcomes of that specific behavior into results that an `Observer` can 
+consume.
+
+To maintain an abstraction between an `Observer`, `Encoder`, and a `Registry`, 
+the return type of a `Registry` should be a wrapper for an `interface{}` type.
+
+```
+type UpkeepResult interface{}
+```
+
+Setting this type specifically allows for distinct types when describing the
+output of one function as the input of another, but provides the necessary type
+abstraction from one registry version to another. Each `Registry` / `Encoder`
+pairing should be responsible for its own type assertions.
+
+## Registry Interfaces
+
+Different `Observer` types can require specific functionality from any 
+particular registry. The folling defines the composability of a `Registry` and
+how that maps to an `Observer` type.
+
+### Must Implement
+
+All registries MUST implement the following interface which performs all checks
+required to determine if an upkeep is eligible to be performed or not. The
+initial intent is to not make a breaking change to the existing registry
+implementation by making a major interface change. The only change here is
+to return the above described type which will just be a wrapper for the existing
+type. An observer does not need to know any of the details of the result. It 
+only needs to be able to pass the result along to an `Encoder`.
+
+```
+type Registry interface {
+    CheckUpkeep(context.Context, ...UpkeepKey) ([]UpkeepResult, error)
+}
+```
+
+### May Implement
```

Comment:
not sure i follow the difference between must and may implement. Do you have any example in mind for a registry which will not implement this?

Reply:
`GetActiveUpkeepIDs` is specific to a polling observer because it's the only case where all upkeep ids are needed at once. A log trigger variant might need some log topic to upkeep id mappings instead or a cron trigger variant might need a cron config to upkeep id mapping.

Reply:
interesting, this needs more thought. Even with log/cron triggers we'll probably need to do balance / pause / cancelled checks so might need this. 

I guess we can explore this later when we design those specific triggers, do you think for now it would affect the refactor design much?